rust-tdshim/src/main.rs

```diff
@@ -156,6 +156,40 @@ const HOB_KERNEL_INFO_GUID: [u8; 16] = [
     0x12, 0xa4, 0x6f, 0xb9, 0x1f, 0x46, 0xe3, 0x4b, 0x8c, 0xd, 0xad, 0x80, 0x5a, 0x49, 0x7a, 0xc0,
 ];
 
+#[repr(u32)]
+#[derive(Clone, Copy, Debug)]
+enum PayloadType {
+    ExecutablePayload,
+    BzImage,
+    RawVmLinux,
+}
+
+fn detect_payload_type(payload: &[u8]) -> PayloadType {
+    let elf64_header = Elf64Header::from_file(payload);
+
+    if elf64_header.e_ident[0] == 0x7f
+        && elf64_header.e_ident[1] == 0x45
+        && elf64_header.e_ident[2] == 0x4c
+        && elf64_header.e_ident[3] == 0x46
+        && elf64_header.e_ident[5] == 0x1
+        && elf64_header.e_phentsize == 0x38
+        && elf64_header.e_phoff >= 0x40
+    {
+        return PayloadType::RawVmLinux;
```

Comment:
Why you use this to detect RawVmLinux?
It is incorrect, because a raw image may strip the elf header.

Reply:
Well the `vmlinux` generated by the kernel build always have the ELF header, that's why I assumed this would be the correct approach. When/How do you get a `vmlinux` without the ELF header?

Reply:
A VMM implementation can do relocation and strip the ELF header.
There is no need to pass the full header, if no load or relocation is required by TdShim.

Reply:
A few questions regarding TdShim:
1. Why does the spec decided to support both `vmlinux` and `bzImage`? `bzImage` has a well defined boot protocol and it should be enough for what TdShim is designed for.
2. Why do you expect the VMM to do relocation and strip the header in the first place? If the spec wasn't accepting this, we would know for sure what `vmlinux` would look like and where to find it.
3. Since the current TdShim implementation doesn't support the `vmlinux` case, don't you think it'd be easier to update the spec and only expect `bzImage`?
4. Is there a QEMU implementation somewhere? I'm asking because I'm curious what the QEMU community thinks about this part of the TdShim spec.

Reply:
Sorry for the slow response and now eventually I reached this part:) @sboeuf @rbradford 
Current there are two ways for the td-shim to load and execute the target linux kernel.
1) The payload area acts as a standard firmware volume and store the target linux kernel as a file in this FV. And [ipl:: find_and_report_entry_point](https://github.com/confidential-containers/td-shim/blob/staging/rust-tdshim/src/ipl.rs#L25) tries to parse the binary. Currently PE32+ and ELF binary are supported.
2) A Kernel Info GUID Hob Entry. When designing the kernel info hob entry, it targets to support raw binary, so it contains an "entrypoint" for td-shim to directly jump to the entry point.

On the other hand, we do have pre-processed binary linux kernel images, so the vmm could directly mmaping the binary file and jump to the entry point. That's why we add the kernel info hob entry to td-shim spec.

Reply:
> On the other hand, we do have pre-processed binary linux kernel images, so the vmm could directly mmaping the binary file and jump to the entry point. That's why we add the kernel info hob entry to td-shim spec.

libkrun does the exact same trick (I wonder who got it first ;-P). The idea is to prepare the kernel image to set it up in a binary state identical to what the kernel expects to have in memory once it reaches the entry point. That's less work for the guest, and also allows directly sharing the kernel image among multiple instances (by mmap'ing the prepared kernel image, as @jiangliu pointed out).

I'm not sure which name should be given to a kernel image like this. `RawVmlinux` seems a bit confusing to me, but perhaps some docs/comments would do the trick.

Reply:
So there are two possible ways forward.
1) extend ipl:: find_and_report_entry_point() to support bzImage.
2) use PayloadInfo::image_type to distinguish pre-processed raw binary or normal linux kernel images.

I'm ok with solution 2 above and hope we could keep the image_type field, so we could have image type for:
1) elf kernel
2) bzImage
3) raw binary
4) auto-detect file type